File: 012PalindromeCheck/palindromeCheck.py
# Solution 1 - iteration
# O(n^2) time / O(n) space

def isPalindrome(string):
    reversedString = ""
    for i in reversed(range(len(string))):
        reversedString += string[i]
    return string == reversedString

# Solution 2 - iteration
# O(n) time / O(n) space


def isPalindrome(string):
    reversedChars = []
    for i in reversed(range(len(string))):
        reversedChars.append(string[i])
    return string == "".join(reversedChars)

# Solution 3 - recursion
# O(n) time / O(n) space
# - check tail recursion for space improvement

def isPalindrome(string, i = 0):
    j = len(string) - 1 - i
    return True if i >= j else string[i] == string[j] and isPalindrome(string, i + 1)

# tail recursion

def isPalindrome(string, i = 0):
    j = len(string) - 1 - i
    if i >= j:
        return True
    if string[i] != string[j]:
        return False
    return isPalindrome(string, i + 1)

# Solution 4
# O(n) time / O(1) space


def isPalindrome(string):
    leftIdx = 0
    rightIdx = len(string) - 1
    while leftIdx < rightIdx:
        if string[leftIdx] != string[rightIdx]:
            return False
        leftIdx += 1
        rightIdx -= 1
    return True

File: 012PalindromeCheck/palindromeCheck.js
// Solution 1 - iteration
// O(n^2) time / O(n) space

const isPalindrome = (string) => {
  let reversedString = '';
  for (let i = string.length - 1; i >= 0; i--) {
    reversedString += string[i];
  }
  return string === reversedString;
};

// Solution 2 - iteration
// O(n) time / O(n) space

const isPalindrome = (string) => {
  let reversedChars = [];
  for (let i = string.length - 1; i >= 0; i--) {
    reversedChars.push(string[i]);
  }
  return string === reversedChars.join('');
};

// Solution 3 - recussion
// O(n) time / O(n) space
// - check tail recursion for space improvement

const isPalindrome = (string, i = 0) => {
  const j = string.length - 1 - i;
  return i >= j ? true : string[i] === string[j] && isPalindrome(string, i + 1);
};

// tail recursion

const isPalindrome = (string, i = 0) => {
    const j = string.length - 1 - i
    if (i >= j) {
        return true
    }
    if (string[i] !== string[j]) {
        return false
    }
    return isPalindrome(string, i + 1)
}

// Solution 4
// O(n) time / O(1) space

const isPalindrome = (string) => {
  let leftIdx = 0;
  rightIdx = string.length - 1;
  while (leftIdx > rightIdx) {
    if (string[leftIdx] !== string[rightIdx]) return false;
    leftIdx++;
    rightIdx--;
  }
  return true;
};
